src/frequenz/channels/_receiver.py

```diff
@@ -342,3 +361,79 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         """Return a string representation of the timer."""
         return f"{type(self).__name__}({self._receiver!r}, {self._mapping_function!r})"
+
+
+class _Sentinel:
+    """A sentinel object to represent no value received yet."""
+
+    def __repr__(self) -> str:
+        """Return a string representation of this sentinel."""
+        return "Sentinel()"
+
+
+_SENTINEL = _Sentinel()
+
+
+class _Filter(Receiver[ReceiverMessageT_co], Generic[ReceiverMessageT_co]):
+    """Apply a filter function on the messages on a receiver."""
+
+    def __init__(
+        self,
+        *,
+        receiver: Receiver[ReceiverMessageT_co],
+        filter_function: Callable[[ReceiverMessageT_co], bool],
+    ) -> None:
+        """Initialize this receiver filter.
+
+        Args:
+            receiver: The input receiver.
+            filter_function: The function to apply on the input data.
+        """
+        self._receiver: Receiver[ReceiverMessageT_co] = receiver
+        """The input receiver."""
+
+        self._filter_function: Callable[[ReceiverMessageT_co], bool] = filter_function
+        """The function to apply on the input data."""
+
+        self._next_message: ReceiverMessageT_co | _Sentinel = _SENTINEL
+
+        self._recv_closed = False
+
+    async def ready(self) -> bool:
+        """Wait until the receiver is ready with a message or an error.
+
+        Once a call to `ready()` has finished, the message should be read with
+        a call to `consume()` (`receive()` or iterated over). The receiver will
+        remain ready (this method will return immediately) until it is
+        consumed.
+
+        Returns:
+            Whether the receiver is still active.
+        """
+        while await self._receiver.ready():
+            message = self._receiver.consume()
+            if self._filter_function(message):
+                self._next_message = message
+                return True
+        self._recv_closed = True
+        return False
+
+    def consume(self) -> ReceiverMessageT_co:
+        """Return a transformed message once `ready()` is complete.
+
+        Returns:
+            The next message that was received.
+
+        Raises:
+            ReceiverStoppedError: If the receiver stopped producing messages.
+            ReceiverError: If there is a problem with the receiver.
+        """
+        if self._recv_closed:
+            raise ReceiverStoppedError(self)
+        assert not isinstance(
+            self._next_message, _Sentinel
+        ), "`consume()` must be preceded by a call to `ready()`"
+
+        message = self._next_message
+        self._next_message = _SENTINEL
+        return message
```

Comment:
It would be good also to add a `__repr__` and `__str__` here too, including the original receiver so one can know the source of this receiver. You can probably copy them from `_Mapper`.